[BUG] certain combinations of $ref overrides cause "ERR! Unable to resolve reference" from arborist

### Is there an existing issue for this?

- [X] I have searched the existing issues

### This issue exists in the latest npm version

- [X] I am using the latest npm

### Current Behavior

If a project has a dependency on `"@types/node": "^16.0.0"` and attempts to use an overrides entry for `"@types/node": "$@types/node"` to pin all references of `@types/node` to v16, and then also attempt to create a dependency on either the `jest` or `jest-cli` packages, arborist becomes confused and produces the following error message:

```
npm ERR! Unable to resolve reference $@types/node
```

(see repro steps below for stack trace from debug logs)

I was unable to work out why this specific combination of packages caused problems. Some of the findings from the unsuccessful theories I investigated include:

* Repros with any package that depends on `jest-config` (including a minimal synthetic package, see repro steps), but does not repro with `jest-config` itself
* No repro with other parent packages of similar complexity with many nested `@types/node` dependencies (eg, `@jest/core`)
* No repro if pinning other `jest-config` dependencies with similar names instead of `@types/node` (eg, `@jest/environment`)
* No repro if the override pins to a version number or version range, only if pinning to a `$ref`
* Repros on both Windows and Linux

### Expected Behavior

npm shouldn't error out; it should install the `jest` dependency correctly, with transitive dependencies on `@types/node` overriden to match the direct `dependency` entry for `@types/node`.

### Steps To Reproduce

```shell
> mkdir npm-cli-overrides-ref-parse-repro && cd npm-cli-overrides-ref-parse-repro && npm init -y
> npm install @types/node@16

# Edit package.json adding: "overrides": { "@types/node": "$@types/node" }

# Motivating case:
> npm install jest # or jest-cli
npm ERR! Unable to resolve reference $@types/node

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\danielbj\AppData\Local\npm-cache\_logs\2022-02-09T23_06_13_180Z-debug-0.log

# More minimal repro (any package wrapping jest-config triggers this, but installing jest-config itself works fine)
> cat ./repro-dependency/package.json
{
    "name": "repro-dependency",
    "description": "repro of Unable to resolve reference npm/cli issue.",
    "version": "27.5.1",
    "main": "./build/index.js",
    "dependencies": {
        "jest-config": "^27.5.1"
    }
}
> cd ./repro-dependency && tar -czvf ./repro-dependency.tar.gz ./package.json && cd ..
> npm install ./repro-dependency/repro-dependency.tar.gz
npm ERR! Unable to resolve reference $@types/node

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\danielbj\AppData\Local\npm-cache\_logs\2022-02-09T23_46_53_856Z-debug-0.log
```

Relevant snippet of [the debug log](https://gist.github.com/dbjorge/75de630ff35797ee268d480e5efbbd95):

```
688 verbose stack Error: Unable to resolve reference $@types/node
688 verbose stack     at Edge.get spec [as spec] (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\node_modules\@npmcli\arborist\lib\edge.js:177:15)
688 verbose stack     at Arborist.[nodeFromEdge] (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\node_modules\@npmcli\arborist\lib\arborist\build-ideal-tree.js:1081:46)
688 verbose stack     at Arborist.[loadPeerSet] (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\node_modules\@npmcli\arborist\lib\arborist\build-ideal-tree.js:1317:36)
688 verbose stack     at async Arborist.[loadPeerSet] (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\node_modules\@npmcli\arborist\lib\arborist\build-ideal-tree.js:1317:11)
688 verbose stack     at async Arborist.[buildDepStep] (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\node_modules\@npmcli\arborist\lib\arborist\build-ideal-tree.js:951:11)
688 verbose stack     at async Arborist.buildIdealTree (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\node_modules\@npmcli\arborist\lib\arborist\build-ideal-tree.js:215:7)
688 verbose stack     at async Promise.all (index 1)
688 verbose stack     at async Arborist.reify (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\node_modules\@npmcli\arborist\lib\arborist\reify.js:152:5)
688 verbose stack     at async Install.exec (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\lib\commands\install.js:157:5)
688 verbose stack     at async module.exports (C:\Users\danielbj\AppData\Local\Volta\tools\image\npm\8.4.1\lib\cli.js:66:5)
689 verbose cwd C:\repos\npm-cli-overrides-ref-parse-repro
690 verbose Windows_NT 10.0.22000
691 verbose argv "C:\\Users\\danielbj\\AppData\\Local\\Volta\\tools\\image\\node\\16.14.0\\node.exe" "C:\\Users\\danielbj\\AppData\\Local\\Volta\\tools\\image\\npm\\8.4.1\\bin\\npm-cli.js" "install" "jest"
692 verbose node v16.14.0
693 verbose npm  v8.4.1
694 error Unable to resolve reference $@types/node
695 verbose exit 1
```


### Environment

- npm: 8.4.1
- Node.js: 16.14.0
- OS Name: Verified on both Windows 11 21H2 22000.493 and also an Ubuntu 20.04 WSL environment (same node/npm versions)
- System Model Name: Surface Laptop 3
- npm config:
```ini
; node bin location = C:\Users\danielbj\AppData\Local\Volta\tools\image\node\16.14.0\node.exe
; cwd = C:\repos\npm-cli-overrides-ref-parse-repro
; HOME = C:\Users\danielbj
; Run `npm config ls -l` to show all defaults.
```

--- a/workspaces/arborist/lib/edge.js
+++ b/workspaces/arborist/lib/edge.js
@@ -169,7 +169,11 @@ class Edge {
     if (this.overrides && this.overrides.value && this.overrides.name === this.name) {
       if (this.overrides.value.startsWith('$')) {
         const ref = this.overrides.value.slice(1)
-        const pkg = this.from.root.package
+        // we may be a virtual root, if we are we want to resolve reference overrides
+        // from the real root, not the virtual one
+        const pkg = this.from.sourceReference
+          ? this.from.sourceReference.root.package
+          : this.from.root.package
         const overrideSpec = (pkg.devDependencies && pkg.devDependencies[ref]) ||
             (pkg.optionalDependencies && pkg.optionalDependencies[ref]) ||
             (pkg.dependencies && pkg.dependencies[ref]) ||

--- a/workspaces/arborist/test/edge.js
+++ b/workspaces/arborist/test/edge.js
@@ -809,6 +809,147 @@ t.same(bundledEdge.explain(), {
   from: bundleParent.explain(),
 }, 'bundled edge.explain as expected')
 
+t.test('override references find the correct root', (t) => {
+  const overrides = new OverrideSet({
+    overrides: {
+      foo: '$foo',
+    },
+  })
+
+  const root = {
+    name: 'root',
+    packageName: 'root',
+    edgesOut: new Map(),
+    edgesIn: new Set(),
+    explain: () => 'root node explanation',
+    package: {
+      name: 'root',
+      version: '1.2.3',
+      dependencies: {
+        foo: '^1.0.0',
+      },
+      overrides: {
+        foo: '$foo',
+      },
+    },
+    get version () {
+      return this.package.version
+    },
+    isTop: true,
+    parent: null,
+    overrides,
+    resolve (n) {
+      return n === 'foo' ? foo : null
+    },
+    addEdgeOut (edge) {
+      this.edgesOut.set(edge.name, edge)
+    },
+    addEdgeIn (edge) {
+      this.edgesIn.add(edge)
+    },
+  }
+
+  const foo = {
+    name: 'foo',
+    packageName: 'foo',
+    edgesOut: new Map(),
+    edgesIn: new Set(),
+    explain: () => 'foo node explanation',
+    package: {
+      name: 'foo',
+      version: '1.2.3',
+      dependencies: {},
+    },
+    get version () {
+      return this.package.version
+    },
+    parent: root,
+    root: root,
+    resolve (n) {
+      return n === 'bar' ? bar : this.parent.resolve(n)
+    },
+    addEdgeOut (edge) {
+      this.edgesOut.set(edge.name, edge)
+    },
+    addEdgeIn (edge) {
+      this.edgesIn.add(edge)
+    },
+  }
+  foo.overrides = overrides.getNodeRule(foo)
+
+  const bar = {
+    name: 'bar',
+    packageName: 'bar',
+    edgesOut: new Map(),
+    edgesIn: new Set(),
+    explain: () => 'bar node explanation',
+    package: {
+      name: 'bar',
+      version: '1.2.3',
+      dependencies: {
+        foo: '^2.0.0',
+      },
+    },
+    get version () {
+      return this.package.version
+    },
+    parent: foo,
+    root: root,
+    resolve (n) {
+      return this.parent.resolve(n)
+    },
+    addEdgeOut (edge) {
+      this.edgesOut.set(edge.name, edge)
+    },
+    addEdgeIn (edge) {
+      this.edgesIn.add(edge)
+    },
+  }
+  bar.overrides = foo.overrides.getNodeRule(bar)
+
+  const virtualBar = {
+    name: 'bar',
+    packageName: 'bar',
+    edgesOut: new Map(),
+    edgesIn: new Set(),
+    explain: () => 'bar node explanation',
+    package: {
+      name: 'bar',
+      version: '1.2.3',
+      dependencies: {
+        foo: '^2.0.0',
+      },
+    },
+    parent: null,
+    root: null,
+    sourceReference: bar,
+    get version () {
+      return this.package.version
+    },
+    resolve (n) {
+      return bar.resolve(n)
+    },
+    addEdgeOut (edge) {
+      this.edgesOut.set(edge.name, edge)
+    },
+    addEdgeIn (edge) {
+      this.edgesIn.add(edge)
+    },
+  }
+  virtualBar.overrides = overrides
+
+  const edge = new Edge({
+    from: virtualBar,
+    type: 'prod',
+    spec: '^2.0.0',
+    name: 'foo',
+    overrides: overrides.getEdgeRule({ name: 'foo', spec: '^2.0.0' }),
+  })
+
+  t.ok(edge.valid, 'edge is valid')
+  t.end()
+})
+
 t.test('shrinkwrapped and bundled deps are not overridden and remain valid', (t) => {
   const overrides = new OverrideSet({
     overrides: {

